nameserver/server.go

```diff
@@ -373,32 +370,42 @@ func (s *DNSServer) localHandler(proto dnsProtocol, kind string, qtype uint16,
 func (s *DNSServer) notUsHandler(proto dnsProtocol) dns.HandlerFunc {
 	dnsClient := proto.GetNewClient(DefaultUDPBuflen, s.timeout)
 
+	localNameResolver := s.localHandler(proto, "Query", dns.TypeA,
+		s.Zone.DomainLookupName, makeAddressReply, s.Zone.ObserveName, failHandleFunc)
+
 	return func(w dns.ResponseWriter, r *dns.Msg) {
 		q := r.Question[0]
-
-		// announce our max payload size as the max payload our client supports
-		maxLen := getMaxReplyLen(r, proto)
-		rcopy := r
-		rcopy.SetEdns0(uint16(maxLen), false)
-
-		Log.Debugf("[dns msgid %d] Fallback query: %+v [%s, max:%d bytes]", rcopy.MsgHdr.Id, q, proto, maxLen)
-		for _, server := range s.Upstream.Servers {
-			reply, _, err := dnsClient.Exchange(rcopy, fmt.Sprintf("%s:%s", server, s.Upstream.Port))
-			if err != nil {
-				Log.Debugf("[dns msgid %d] Network error trying %s (%s)",
-					r.MsgHdr.Id, server, err)
-				continue
-			}
-			if reply != nil && reply.Rcode != dns.RcodeSuccess {
-				Log.Debugf("[dns msgid %d] Failure reported by %s for query %s",
+		nc := nameNumComponents(q.Name)
+		if nc == 1 {
+			Log.Debugf("[dns msgid %d] Unqualified query: '%+v' ", r.MsgHdr.Id, q)
+			localNameResolver(w, r) // short names (eg, "redis.") will follow a local resolution process
+			return
+		} else if nc > 1 {
+			// announce our max payload size as the max payload our client supports
+			maxLen := getMaxReplyLen(r, proto)
+			rcopy := r
+			rcopy.SetEdns0(uint16(maxLen), false)
+
+			Log.Debugf("[dns msgid %d] Fallback query: %+v [%s, max:%d bytes]", rcopy.MsgHdr.Id, q, proto, maxLen)
+			for _, server := range s.Upstream.Servers {
+				reply, _, err := dnsClient.Exchange(rcopy, fmt.Sprintf("%s:%s", server, s.Upstream.Port))
+				if err != nil {
+					Log.Debugf("[dns msgid %d] Network error trying %s (%s)",
+						r.MsgHdr.Id, server, err)
+					continue
+				}
+				if reply != nil && reply.Rcode != dns.RcodeSuccess {
+					Log.Debugf("[dns msgid %d] Failure reported by %s for query %s",
+						r.MsgHdr.Id, server, q.Name)
+					continue
+				}
+				Log.Debugf("[dns msgid %d] Given answer by %s for query %s",
 					r.MsgHdr.Id, server, q.Name)
-				continue
+				w.WriteMsg(reply)
+				return
```

Comment:
Because, once we have spent some CPU cycles calculating the number of components in the name, we could use that number for skipping empty names...